### PR TITLE
refactor core to use constructs and make it compatible with compiler

### PR DESCRIPTION
**Pull request by jhsinger-klotho**
• Does any part of it require special attention?
• Does it relate to or fix any issue?

This is the start of removing CloudResource/CompilationResult and what that means today and moving in the new structs/interfaces we are talking about.

This PR:

 - Makes all of our resources conform to the new Construct interface
 -  Splits out resources to be kind based (Really only affects persist since it was wrapping multiple kinds)
 - Makes config not care about kind but rather care about struct type
 - Creates compiler and compilation document outside of core so we dont have circular deps
 - Makes EnvironmentVariables attach the construct so we dont need kind

Removals:
- Infra as code (with outputFiles, thats where the files from IaC Generation should go now)
-  dependencies (these are represented in the graph itself now)
- Topology - For now i remove topology since it will break even more things and we will add it back at the end

This pr is very much breaking everything, but im going to push up to different branches one by one until i have the entirety of the compilation portion working

### Standard checks

- **Unit tests**: Any special considerations?
- **Docs**: Do we need to update any docs, internal or public?
- **Backwards compatibility**: Will this break existing apps? If so, what would be the extra work required to keep them working?

### CODE REVIEW

File: pkg/core/resources.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import "github.com/dominikbraun/graph"
+
+type (
+	Construct interface {
+		Provenance() AnnotationKey
+		Id() string
+	}
+
+	CloudResource interface {
```
Comment (gordon-klotho) [inline pkg/core/resources.go:11]: Can we shorten this to `Resource`? Maybe even `Infrastructure` (if we're open to renaming). Since we're determining the new name and I wouldn't want to change it often, I think supporting on-prem and metal should indicate that this shouldn't be "cloud" specific (also why "provider" is named such).

File: pkg/compiler/compilation_document.go [hunk 2]
```diff
@@ -4,23 +4,26 @@ import (
...
 )
 
 type (
 	CompilationDocument struct {
-		InputFiles     *InputFiles
-		Constructs     graph.Directed[Construct]
-		CloudResources graph.Directed[ProviderResource]
-		OutputFiles    []File
+		InputFiles     *core.InputFiles
+		Constructs     *graph.Directed[core.Construct]
```
Comment (gordon-klotho) [inline pkg/compiler/compilation_document.go:15]: I'd recommend doing a
```go
type ConstructGraph = graph.Directed[core.Construct]
```
Saves on some typing and redundant info.

File: pkg/compiler/compilation_document.go [hunk 3]
```diff
@@ -4,23 +4,26 @@ import (
...
+		Configuration  *config.Application
+		CloudResources []*graph.Directed[core.CloudResource]
```
[the remark this answers is not in the document]
Comment (jhsinger-klotho) [reply pkg/compiler/compilation_document.go:17]: if we have multiple providers, how would one know how to input? Or are we saying they would just be more root nodes?

File: pkg/compiler/compiler.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+type (
+	Plugin interface {
+		Name() string
+	}
+
+	AnalysisAndTransformationPlugin interface {
+		Name() string
+
+		// Transform is expected to mutate the result and any dependencies
+		Transform(*core.InputFiles, *graph.Directed[core.Construct]) error
```
[the remark this answers is not in the document]
Comment (jhsinger-klotho) [reply pkg/compiler/compiler.go:21]: becuase these all run on the same graph, we want it to centrally mutate that graph. If providers done return their own DAG then we need them to mutate a central one as well.

File: pkg/core/resources.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package core
+
+import "github.com/dominikbraun/graph"
+
+type (
+	Construct interface {
```
[the remark this answers is not in the document]
Comment (jhsinger-klotho) [reply pkg/core/resources.go:6]: where do you suggest it goes? I didnt want to make a file just for the type and i add to this with functionality later on

File: pkg/compiler/compilation_document.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gordon-klotho) [reply pkg/compiler/compilation_document.go:17]: Yeah, the graph could represent multiple subgraphs which could be independent.

[the remark this answers is not in the document]
Comment (jhsinger-klotho) [reply pkg/compiler/compilation_document.go:17]: ok ill modify this as well then. We then need to pass it into the plugins so its mutable rather than it being returned

File: pkg/compiler/compiler.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+func (c *Compiler) Compile() error {
+	// Add our internal resource to be used for provider specific implementations. ex) aws dispatcher requires the payloads bucket and so does proxy
+	// TODO: We could likely move this into runtime, but until we refactor that to be common we can keep this here so it lives in one place.
+	// We previously always created the payloads bucket so the behavior is no different
+	internalResource := &core.InternalResource{AnnotationKey: core.AnnotationKey{ID: core.KlothoPayloadName, Capability: annotation.InternalCapability}}
+	c.Document.Constructs.AddVertex(internalResource)
+
+	for _, p := range c.AnalysisAndTransformationPlugins {
```
[the remark this answers is not in the document]
Comment (jhsinger-klotho) [reply pkg/compiler/compiler.go:57]: right but the provider plugin is useless without analysis and transformations having run and the IaC plugin is useless without the provider plugins having run, so i dont see this as being more enforced that it was previously

File: pkg/compiler/compilation_document.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gordon-klotho) [reply pkg/compiler/compilation_document.go:17]: Yeah, and I think that's better as it allows for cross-provider things as well.

File: pkg/core/resources.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gordon-klotho) [reply pkg/core/resources.go:9]: Probably compilation_document if it's going to be too small for a file on its own.

File: pkg/lang/javascript/expose.go [hunk 7]
```diff
@@ -110,7 +109,7 @@ func handleGatewayRoutes(info *execUnitExposeInfo, constructGraph *graph.Directe
 				// TODO when there are ways to combine units, we'll need a more sophisticated way to see which unit the target maps to.
 				depKey.ID = info.Unit.ID
 			}
-			constructGraph.AddEdge(gw.Provenance().ToString(), depKey.ToString())
+			constructGraph.AddEdge(gw.Id(), depKey.ToString())
```
Comment (gordon-klotho) [inline pkg/lang/javascript/expose.go:112]: Ah, now I see why `ToString` was added. Usually I'd associate the `ToString` with textual representation meant for humans (it's the `Stringer` interfaced and used in `fmt.Print` functions). I have a slight preference to name this more like `ToId`.

**Review by gordon-klotho — COMMENTED** [review]
22b1e79a590abe1927edc38a405e4823ae985d3d 👍

File: pkg/lang/javascript/expose.go (hunk 7 above)
Comment (jhsinger-klotho) [reply pkg/lang/javascript/expose.go:112]: Thats fair, i think we can rename that

Comment (jhsinger-klotho) [reply pkg/lang/javascript/expose.go]: renamed in the most recent commint

File: pkg/core/resources.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhsinger-klotho) [reply pkg/core/resources.go:9]: then were back to having a circular dependency. They need to stay in the core package since config uses them and the compiler pkg uses config. 

Im fine putting them elsewhere but it needs to be within this package
